### PR TITLE
Python: Share Python interpreter across requests

### PR DESCRIPTION
**Pull request by hoodmane**
Before this PR we instantiate the Python interpreter from scratch on every request. This is slow and uses up a lot of memory. This switches to sharing an interpreter across requests. This makes per-request non-coldstart execution time and memory usage much lower.

### CODE REVIEW

File: src/pyodide/python-entrypoint-helper.js [hunk 1]
```diff
@@ -210,16 +210,26 @@ async function setupPackages(pyodide) {
   return pyodide.pyimport(mainModule);
 }
 
+let mainModule;
+async function getMainModule() {
+  if (mainModule) {
```
Comment (jasnell) [inline src/pyodide/python-entrypoint-helper.js:215]:
```suggestion
  if (mainModule === undefined) {
```

Comment (jasnell) [reply src/pyodide/python-entrypoint-helper.js:215]: If this method ends up getting called multiple times without `await` we'll end up with multiple calls to `loadPyodide` running. If the call is already inflight, in would be good to just attach to the existing promise.

Comment (hoodmane) [reply src/pyodide/python-entrypoint-helper.js]: How's it look now?

File: src/pyodide/python-entrypoint-helper.js [hunk 2]
```diff
@@ -210,16 +210,28 @@ async function setupPackages(pyodide) {
...
+  if (mainModulePromise !== undefined) {
+    return mainModulePromise;
+  }
+  mainModulePromise = (async function() {
+    // TODO: investigate whether it is possible to run part of loadPyodide in top level scope
+    // When we do it in top level scope we seem to get a broken file system.
+    const pyodide = await loadPyodide();
+    return await setupPackages(pyodide);
+  })();
+  return mainModulePromise;
```
Comment (jasnell) [inline src/pyodide/python-entrypoint-helper.js:224]: Bit easier with at least one fewer Promise allocated...

```suggestion
  mainModulePromise = setupPackages(await loadPyodide());
  return mainModulePromise;
```

Comment (hoodmane) [reply src/pyodide/python-entrypoint-helper.js:224]: Is this still guaranteed to only execute once? As I understand it, 
```js
mainModulePromise = setupPackages(await loadPyodide());
```
means the same thing as:
```js
const tmp = await loadPyodide();
mainModulePromise = setupPackages(tmp);
```
and so there is an await before assigning to `mainModulePromise` meaning that a second request could come in and observe `mainModulePromise` as not null. The way I wrote it with the outside function being synchronous and assigning to the promise guarantees that we don't yield to the event loop when `mainModulePromise` is still `undefined.

Comment (jasnell) [reply src/pyodide/python-entrypoint-helper.js:224]: hah... doh you're right lol. that's what I get for doing code review at dinner time.

**Review by jasnell — APPROVED** [review]

**Review by dom96 — APPROVED** [review]

**Comment by hoodmane** [comment]
My concern with rebasing is that I thought github would delete review comments on the old commits? It's fine if you've read and responded to them as long as nobody but it makes the PR history difficult to understand and you can lose comments if you force push before reading the PR. Or is this not right? I'm not really sure what the other tradeoffs are with rebase to understand under which circumstances one or the other is preferred. With stacked PRs merge is really annoying, but stacked PRs are really annoying in general.

**Comment by dom96** [comment]
> My concern with rebasing is that I thought github would delete review comments on the old commits?

I don't believe that happens

**Comment by hoodmane** [comment]
Okay, probably I should just try it for a while and see. Thanks!

**Pull request merged**